### DISCUSSION TITLE
How to use collation?

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#1946] [Q&A] [answered]
Is there a way to do the equivalent of this:

https://stackoverflow.com/questions/2461522/how-do-i-perform-an-accent-insensitive-compare-e-with-%C3%A8-%C3%A9-%C3%AA-and-%C3%AB-in-sql-ser

in order to perform accent insensitive queries?

```
(select(appUser)
            ..limit(limit)
            ..where((c) =>
                c.firstName.lower().contains(query.toLowerCase()) |
                c.lastName.lower().contains(query.toLowerCase()) |
                c.username.lower().contains(query.toLowerCase())))
          .get()
```

**Comment by simolus3** [comment, marked as answer]
You can use `c.firstName.collate('collation')` in drift's Dart API to use the `COLLATE` operator. However, be aware that sqlite3 [only has three builtin collations](https://www.sqlite.org/datatype3.html#collation) and none of them are suitable for case-insensitive text comparisons outside of regular ASCII chars (the same goes for `lower()` and `upper`() btw, they don't do anything for most chars).

If you need support for this that works across locales, you probably need to include an ICU library and write custom sqlite3 functions or collations for that. I can help you with the later, but I don't know a good ICU library for Dart.

**Reply by BananaMasterz** [reply]
Oh I see... do you think there's another workaround other than custom sqlite3 functions? For now I'll just have a new column in my table with an accent-free and uppercase value of the original to compare it against. It will increase the overall size of the database, but it's something readily available.... If I think of anything better in the meantime or if you come up with something that works, I'll switch to that.

Thank you for your work and response sir

**Reply by simolus3** [reply]
Do you know if there's a way to remove accents in Dart? I'm not too familiar with that and I think there's nothing in the SDK for it, but it may be that something like this exists for Flutter? If it can be written in Dart, I can help you get it working in SQL as well :D

**Reply by BananaMasterz** [reply]
haven't found anything myself other than mapping the characters like this:

https://stackoverflow.com/questions/30844353/how-to-remove-diacritics-accents-from-a-string

but for the Greek alphabet (since the app I'm building is in Greek only)

**Reply by simolus3** [reply]
So you can create a custom collation by registering it to a `NativeDatabase` in the `setup` parameter, like this:

```dart
NativeDatabase.memory(setup: (rawDb) {
  rawDb.createCollation(
      name: 'latin',
      function: (a, b) {
        if (a == null) return -1;
        if (b == null) return 1;

        return removeDiacritics(a).compareTo(removeDiacritics(b));
      });
});
```

In the next drift version, you will be able do use something like `table.textColumn.collate(Collation('latin'))` to use this custom collation.

At the moment, you unfortunately have to use `table.textColumn.collateCustom('latin')` with this monstrosity:

```dart
extension MyCollate on Expression<String?> {
  Expression<String?> collateCustom(String collation) =>
      _CollateOperator(this, collation);
}

class _CollateOperator extends Expression<String> {
  final Expression inner;
  final String collate;

  @override
  final Precedence precedence = Precedence.postfix;

  _CollateOperator(this.inner, this.collate);

  @override
  void writeInto(GenerationContext context) {
    writeInner(context, inner);
    context.buffer
      ..write(' COLLATE ')
      ..write(collate);
  }

  @override
  int get hashCode => Object.hash(inner, collate);

  @override
  bool operator ==(Object other) {
    return other is _CollateOperator &&
        other.inner == inner &&
        other.collate == collate;
  }
}
```

**Reply by BananaMasterz** [reply]
Well if it works it's way less of a monstrosity than what I've done already haha, will test. Huge thanks!

**Reply by BananaMasterz** [reply]
Are you going to use this package?

https://pub.dev/packages/diacritic

cuz this doesn't work for greek accents I just tried with ά should turn into A but it does nothing. I saw the map used here:
https://github.com/agilord/diacritic/blob/master/lib/src/replacement_map.dart
but it's very limited :(

**Reply by BananaMasterz** [reply]
So I added the greek alphabet to the map linked above and it works for converting 'έ' to 'e' for example.

Now what I want to do is 

say we have the column `username` with `νικ` in it.
I want here

`..where((c) => c.username.collateCustom('latin').contains(query)))`

to pass the `query='νίκ'` and get the row of  `νικ` returned but it doesn't work.... 

what DOES work tho, is having `nik` in the column and passing `νίκ` in the query... I'm not sure how to reverse this in the code you send me...

**Reply by North101** [reply]
Does this work?
```dart
..where((c) => c.username.contains(query) | c.username.collateCustom('latin').contains(removeDiacritics(query))))
```

**Reply by BananaMasterz** [reply]
it doesn't... same issue... returns `νικ` when I type `nik` but when I type `νίκ` it doesn't return anything

**Reply by simolus3** [reply]
The problem is that, in drift, `contains(x)` translates to `LIKE %x%`. In sqlite3, the `LIKE` operator does not use a collation as it seems. It always compares utf16 code units (with an optional mode to allow case-insensitive comparisons for ASCII characters).

If you already have your `removeDiacritics()` function with support for greek characters, you could register a custom sql function for that as well:

```dart
  rawDb.createFunction(
    functionName: 'removeDiacritics',
    function: (args) {
      return removeDiacritics(args[0] as String);
    },
    argumentCount: const AllowedArgumentCount(1),
  );
```

Then, you can define an extension like this:

```dart
extension MyCollate on Expression<String?> {
  Expression<String> removeDiacritics() => FunctionCallExpression('removeDiacritics', [this]);

  Expression<String?> collateCustom(String collation) =>
      _CollateOperator(this, collation);
}
```

Then, `c.username.removeDiacritics().contains(removeDiacritics(query))` should work since it removes diacritics on both sides of the `LIKE` operator.

So to summarize:

- you register a custom collation to compare (`==`, `!=`, `<`, `<=`, `>`, `>=`) two strings without diacritics
- for more complex queries including `LIKE` searches with `contains()`, you register a custom functions to remove diacritics from a term entirely.

**Reply by BananaMasterz** [reply]
It's amazing how customizable drift is. I'll try although I have 5 days experience in drift but your reply seems easy enough to follow. Thanks

**Reply by BananaMasterz** [reply]
so I tried 
```dart
rawDb.createFunction(
  functionName: 'removeDiacritics',
  function: (args) {
    return removeDiacritics(args[0] as String);
  },
  argumentCount: AllowedArgumentCount(1),
);
````

with this

```dart
extension MyCollate on Expression<String?> {
  Expression<String> removeDiacritics() => FunctionCallExpression('removeDiacritics', [this]);

  Expression<String?> collateCustom(String collation) => _CollateOperator(this, collation);
}


class _CollateOperator extends Expression<String> {
  final Expression inner;
  final String collate;

  @override
  final Precedence precedence = Precedence.or;

  _CollateOperator(this.inner, this.collate);

  @override
  void writeInto(GenerationContext context) {
    writeInner(context, inner);
    context.buffer
      ..write(' COLLATE ')
      ..write(collate);
  }

  @override
  int get hashCode => Object.hash(inner, collate);

  @override
  bool operator ==(Object other) {
    return other is _CollateOperator && other.inner == inner && other.collate == collate;
  }
}
```

and no matter what, I cannot get it to work... I pass `τεστ` and the database column has `τέστ`  and it doesn't match it... the opposite works tho (pass `τέστ` when db has `τεστ`) which means the `.collateCustom('removeDiacritics')` isn't doing anything. Not sure what I'm doing wrong... I tried many combos too such as:

```dart
c.username.lower().collateCustom('removeDiacritics').contains(removeDiacritics(query)) |
c.username.lower().collateCustom('removeDiacritics').like(query) |
c.username.lower().contains(query.toLowerCase()) |
c.username.lower().contains(query) |
c.username.upper().contains(query.toUpperCase()) |
c.username.collateCustom('removeDiacritics').lower().contains(removeDiacritics(query))
```

**Reply by North101** [reply]
As @simolus3 said, you only use `collateCustom` when checking for equality or order. When you want to use special functions like `compare` or `like`, instead use the `removeDiacritics` function on the column.

E.g.
```
c.username.removeDiacritics().lower().contains(removeDiacritics(query).toLowerCase())
```

**Reply by BananaMasterz** [reply]
oh god.... I think I spent too much time without a break... you're absolutely right. It works now. I don't know whether to mark something in this discussion as an answer or not since it's a custom solution?

**Reply by North101** [reply]
I think `drift` wants to be unopinionated about this sort of thing. It only wants to be a nice way of interacting with sqlite through `dart`. Any additional functionality like syncing between a server and client or comparing text ignoring accents etc fall outside of what drift wants to do, though it does give you the tools to do that yourself which is what this solution is.

**Comment by North101** [comment]
FYI: For nullable columns I had to change the function to:
```dart
rawDb.createFunction(
  functionName: 'removeDiacritics',
  function: (args) {
    final text = args[0] as String?;
    return text != null ? removeDiacritics(text) : null;
  },
),
```

**Comment by BananaMasterz** [comment]
@simolus3 I'm trying the following:

```dart
var x = await (db.select(db.contactCarrier)
      ..where((t) => t.carrier.removeDiacritics().contains(removeDiacritics('ΤΤ'))))
  .get();
  ```

and I have a single row with `ΤΤ` in it and it's not matching it, but if i do a single `T` instead of 2, it matches.

```dart
var x = await (db.select(db.contactCarrier)
      ..where((t) => t.carrier.removeDiacritics().contains(removeDiacritics('Τ')))) // single T instead of 2
  .get();
  ```

if I change my row to have `TT a` or a space and any character after the double T, then the first also matches... 

what's weird is that the in this:

```dart
rawDb.createFunction(
  functionName: 'removeDiacritics',
  function: (args) {
    final text = args[0] as String?;
    print('${text == null ? 'null' : removeDiacritics(text).contains(removeDiacritics('ΤΤ'))}');
    return text != null ? removeDiacritics(text) : null;
  },
);
```

it prints true so shouldn't it be returned in the query? (the removeDiacritics is modified for greek characters but I don't think it matters since it works otherwise)

I first thought it may be an issue with `contains` not matching if it's the exact same string but i tried with equals as well and the issue still came up, so I have no idea how can the `rawDb.createFunction` print true but not have it returned from the query


EDIT: I think there's an issue with `collate` because the same code used to work some time ago (not sure with what drift version)

**Reply by simolus3** [reply]
Thanks for the report!

> EDIT: I think there's an issue with `collate` because the same code used to work some time ago (not sure with what drift version)

Are you using sqlite3 collations here? There was a recent change in the `sqlite3` package that could have introduced this bug, but it seems like you're only using custom functions right? Do you have the original value in the database or a small repro for me to take a look at?

**Reply by BananaMasterz** [reply]
sure, here's an example:

This is my database declaration (in this example I'm not even doing anything to the string since just using the functions seems to trigger it so for simplification I just return the string as is):
```dart
return NativeDatabase(file, setup: (rawDb) {
  rawDb.createFunction(
    functionName: 'removeDiacritics',
    function: (args) {
      final text = args[0] as String?;
      return text != null ? (text) : null;
    },
  );
});
```
And then I have a table `contact` that has a column `firstName` and I run the following:

```dart
var db = Provider.of<MyDatabase>(context, listen: false);
var query = 'ΑΤΤΙ';
// var sqlQuery = db.select(db.contact)
//   ..where((tbl) => tbl.firstName.contains(query));
var sqlQuery = db.select(db.contact)
  ..where((tbl) => tbl.firstName.removeDiacritics().contains(query));
var result = await sqlQuery.get();
print(result);
```
The value in the database is exactly `Ν.Π.Δ.Δ ΠΕΡΙΦΕΡΕΙΑ ΑΤΤΙΚΗΣ`. The above code prints an empty list, but if you uncomment the commented sqlQuery, it prints the row as expected. ( Btw ensure you copy paste, since `ΑΤΤΙ` are greek characters that look identical to english characters.)

Forgot the extension:

```dart
extension MyCollate on Expression<Object> {
  Expression<String> removeDiacritics() => FunctionCallExpression('removeDiacritics', [this]);

  Expression<Object> collateCustom(String collation) => _CollateOperator(this, collation);
}

class _CollateOperator extends Expression<String> {
  final Expression inner;
  final String collate;

  @override
  final Precedence precedence = Precedence.or;

  _CollateOperator(this.inner, this.collate);

  @override
  void writeInto(GenerationContext context) {
    writeInner(context, inner);
    context.buffer
      ..write(' COLLATE ')
      ..write(collate);
  }

  @override
  int get hashCode => Object.hash(inner, collate);

  @override
  bool operator ==(Object other) {
    return other is _CollateOperator && other.inner == inner && other.collate == collate;
  }
}
```

**Reply by simolus3** [reply]
You're right, this was a bug in the `sqlite3` Dart package that I've now fixed. Upgrading dependencies (making sure that `sqlite3` is at least at `1.11.1`) should resolve the issue.

**Reply by BananaMasterz** [reply]
Amazing, thanks! What was the issue?

**Reply by simolus3** [reply]
[Me being foolish](https://github.com/simolus3/sqlite3.dart/commit/9494c31d98366b24eb34ffbcf7444c19af1c565f#diff-017481f7b51e619897f4433de270fa03bf95cce03d1b8acae596fd8db1e66847) (the bug was introduced after adding an abstraction layer to share more code between the `dart:ffi` variant and one that uses `dart:js` in the browser). When the utf8 encoding used to have more bytes than there were characters in the string, we'd cut off text at the end.
